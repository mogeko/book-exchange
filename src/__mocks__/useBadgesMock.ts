import * as hooks from "@/lib/hooks/useBadges";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "default");

const useBadgeMock = {
  target: hooks.default,
  returnResult: () => mock.mockImplementation(() => exampleData),
};

export const exampleData: BadgesType = {
  menus: Array.from({ length: 3 }, () => faker.datatype.number(100)),
  user: Array.from({ length: 2 }, () => faker.datatype.number(100)),
};

type BadgesType = ReturnType<typeof hooks.default>;

export default useBadgeMock;
