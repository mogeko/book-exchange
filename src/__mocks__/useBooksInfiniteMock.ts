import * as hooks from "@/lib/hooks/useBooks";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "useBooksInfinite");
const setSizeMock = jest.fn();

const useBooksInfiniteMock = {
  target: hooks.useBooksInfinite,
  returnResult: (pages = 3) => ({
    success: () =>
      mock.mockImplementation((param) =>
        genExampleRes({}, [pages, param!.limit!])
      ),
    error: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ data: undefined, isError: true }, [
          pages,
          param!.limit!,
        ])
      ),
    loading: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ data: undefined, isLoading: true }, [
          pages,
          param!.limit!,
        ])
      ),
  }),
  genExampleRes,
  setSizeMock,
};

export const exampleData: BookTypes = {
  title: faker.word.noun(20),
  cover: faker.image.image(1280, 1910),
  description: faker.lorem.paragraph(10),
  published: faker.date.past().toISOString(),
  publisher: faker.company.companyName(),
  tags: Array.from({ length: faker.datatype.number({ min: 2, max: 8 }) }, () =>
    faker.lorem.words(2)
  ),
  author: faker.name.firstName(),
  rates: faker.datatype.number(100),
  id: faker.datatype.number(100),
};

function genExampleRes(res: Partial<ResType> = {}, [x, y] = [3, 10]): ResType {
  const exampleRes = {
    data: Array.from({ length: x }, () =>
      Array.from({ length: y }, () => exampleData)
    ),
    isValidating: false,
    isError: false,
    isLoading: false,
    mutate: jest.fn(),
    setSize: setSizeMock,
    size: 1,
  };

  return { ...exampleRes, ...res };
}

type ResType = ReturnType<typeof hooks.useBooksInfinite>;
type BookTypes = hooks.BookTypes;

export default useBooksInfiniteMock;
