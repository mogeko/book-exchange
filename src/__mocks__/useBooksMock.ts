import * as hooks from "@/lib/hooks/useBooks";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "default");

const useBooksMock = {
  target: hooks.default,
  returnResult: () => ({
    success: () =>
      mock.mockImplementation((param) => genExampleRes({}, param!.limit)),
    error: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ books: undefined, isError: true }, param!.limit)
      ),
    loading: () =>
      mock.mockImplementation((param) =>
        genExampleRes({ books: undefined, isLoading: true }, param!.limit)
      ),
  }),
  genExampleRes,
};

export const exampleData: BookType = {
  id: `bk${faker.datatype.number(100)}`,
  title: faker.word.noun(20),
  cover: faker.image.image(1280, 1910),
  tags: Array.from({ length: faker.datatype.number({ min: 2, max: 8 }) }, () =>
    faker.lorem.words(2)
  ),
  rates: faker.datatype.number(100),
  mate: {
    author: faker.name.firstName(),
  },
  desc: {
    short_desc: faker.lorem.paragraph(10),
  },
};

function genExampleRes(res: Partial<ResType> = {}, limit = 10): ResType {
  const exampleRes = {
    books: Array.from({ length: limit }, () => exampleData),
    isValidating: false,
    isError: false,
    isLoading: false,
    mutate: jest.fn(),
  };

  return { ...exampleRes, ...res };
}

type ResType = ReturnType<typeof hooks.default>;
type BookType = hooks.BooksType[0];

export default useBooksMock;
