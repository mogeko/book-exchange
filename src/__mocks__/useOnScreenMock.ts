import * as hooks from "@/lib/useOnScreen";

const useOnScreenMock = {
  target: hooks.default,
  visiable: () => jest.spyOn(hooks, "default").mockImplementation(() => true),
  not: {
    visiable: () =>
      jest.spyOn(hooks, "default").mockImplementation(() => false),
  },
};

export default useOnScreenMock;
