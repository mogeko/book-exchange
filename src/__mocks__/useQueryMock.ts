import * as hooks from "@/lib/hooks/useQuery";

const mock = jest.spyOn(hooks, "default");

const useQueryMock = {
  target: hooks.default,
  returnResult: (res = {}) => mock.mockImplementation(() => genExampleRes(res)),
};

export function genExampleRes(res: Partial<ResType> = {}) {
  const exampleRes = {
    data: undefined,
    isValidating: false,
    isError: false,
    isLoading: false,
    mutate: jest.fn(),
  };

  return { ...exampleRes, ...res };
}

type ResType = hooks.QueryReturn;

export default useQueryMock;
