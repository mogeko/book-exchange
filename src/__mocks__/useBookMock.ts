import * as hooks from "@/lib/hooks/useBooks";
import { exampleData as exampleBooksData } from "@/__mocks__/useBooksMock";
import { genExampleRes } from "@/__mocks__/useQueryMock";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "useBook");

const useBookMock = {
  target: hooks.useBook,
  returnResult: () => ({
    success: () =>
      mock.mockImplementation(() => genExampleRes({ data: exampleData })),
    error: () =>
      mock.mockImplementation(() => genExampleRes({ isError: true })),
    loading: () =>
      mock.mockImplementation(() => genExampleRes({ isLoading: true })),
  }),
};

export const exampleData: BookType = {
  ...exampleBooksData,
  mate: {
    ...exampleBooksData.mate,
    publisher: faker.company.companyName(),
    subtitle: faker.lorem.sentence(10),
    language: "English",
    publication_date: faker.date.past().toISOString(),
    isbn: `978-${faker.datatype.number({ min: 1000000000, max: 1999999999 })}`,
    paperback: faker.datatype.number({ min: 100, max: 1000 }),
  },
  desc: {
    text: faker.lorem.paragraph(50),
    is_folded: faker.datatype.boolean(),
  },
  digest: {
    text: faker.lorem.paragraph(50),
    is_folded: faker.datatype.boolean(),
  },
};

type BookType = hooks.BookType;

export default useBookMock;
