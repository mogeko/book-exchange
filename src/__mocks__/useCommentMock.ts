import * as hooks from "@/lib/hooks/useComments";
import { genExampleRes } from "@/__mocks__/useQueryMock";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "default");

const useCommentMock = {
  target: hooks.default,
  returnResult: () => ({
    success: () =>
      mock.mockImplementation(() => genExampleRes({ data: exampleData })),
    error: () =>
      mock.mockImplementation(() => genExampleRes({ isError: true })),
    loading: () =>
      mock.mockImplementation(() => genExampleRes({ isLoading: true })),
  }),
};

export const exampleData: CommentType = {
  cm_id: `cm${faker.datatype.number({ min: 1000000000, max: 100000000000 })}`,
  msg: faker.lorem.paragraph(50),
};

type CommentType = hooks.CommentType;

export default useCommentMock;
