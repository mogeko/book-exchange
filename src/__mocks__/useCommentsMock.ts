import * as hooks from "@/lib/hooks/useComments";
import { genExampleRes } from "@/__mocks__/useQueryMock";
import { faker } from "@faker-js/faker";

const mock = jest.spyOn(hooks, "default");

const useCommentsMock = {
  target: hooks.default,
  returnResult: () => ({
    success: () =>
      mock.mockImplementation((param) =>
        genExampleRes({
          data: Array.from({ length: param?.limit! }, () => exampleData),
        })
      ),
    error: () =>
      mock.mockImplementation(() => genExampleRes({ isError: true })),
    loading: () =>
      mock.mockImplementation(() => genExampleRes({ isLoading: true })),
  }),
};

export const exampleData: CommentsType = {
  id: `cm${faker.datatype.number({ min: 1000000000, max: 100000000000 })}`,
  author_meta: {
    uid: `${faker.datatype.number({ min: 10000, max: 100000 })}`,
    username: faker.internet.userName(),
    email: faker.internet.email(),
    avatar: faker.internet.avatar(),
  },
  meta: {
    short_review: faker.lorem.lines(1),
    rates: faker.datatype.number(100),
    likes: faker.datatype.number(1000),
    dislike: faker.datatype.number(1000),
    created_at: faker.date.past().toISOString(),
    location: faker.address.county(),
  },
  responds: Array.from(
    { length: faker.datatype.number({ min: 1, max: 10 }) },
    () => ({
      id: `cm${faker.datatype.number({ min: 1000000000, max: 100000000000 })}`,
      author_meta: {
        uid: `${faker.datatype.number({ min: 10000, max: 100000 })}`,
        username: faker.internet.userName(),
        email: faker.internet.email(),
        avatar: faker.internet.avatar(),
      },
      meta: {
        likes: faker.datatype.number(1000),
        dislike: faker.datatype.number(1000),
        created_at: faker.date.past().toISOString(),
        location: faker.address.county(),
      },
      belongs_to: `cm${faker.datatype.number({
        min: 1000000000,
        max: 100000000000,
      })}`,
      msg: faker.lorem.lines(1),
      is_folded: faker.datatype.boolean(),
    })
  ),
  belongs_to: `bk${faker.datatype.number(1000)}`,
  msg: faker.lorem.sentence(),
  is_folded: faker.datatype.boolean(),
};

type CommentsType = hooks.CommentsType[0];

export default useCommentsMock;
