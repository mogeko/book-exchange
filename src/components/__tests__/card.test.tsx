import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { exampleData } from "@/__mocks__/useBooksMock";
import Card, { LongCard } from "@/components/card";

describe("Card", () => {
  it("renders a book card correctly", () => {
    const { container } = render(<Card {...exampleData} />);

    expect(container.querySelector("figure a")).toHaveAttribute(
      "href",
      `/books/${exampleData.id}`
    );
    expect(container.querySelector("figure img")).toBeInTheDocument();
    expect(container.querySelector("h2")?.textContent).toBe(exampleData.title);
    expect(container.querySelector("p")?.textContent).toBe(exampleData.author);
  });

  it("snapshot a book card", () => {
    const { container } = render(<Card {...exampleData} />);

    expect(container).toMatchSnapshot();
  });

  it("snapshot a book skeleton card", () => {
    const { container } = render(<Card.Skeleton />);

    expect(container).toMatchSnapshot();
  });

  it("render a long card", () => {
    const { container } = render(<LongCard {...exampleData} />);

    expect(container.querySelector("figure a")).toHaveAttribute(
      "href",
      `/books/${exampleData.id}`
    );
    expect(container.querySelector("figure img")).toBeInTheDocument();
    expect(container.querySelector("h2")?.textContent).toBe(exampleData.title);
    expect(container.querySelector("p")?.textContent).toBe(exampleData.author);
  });

  it("snapshot a long card", () => {
    const { container } = render(<LongCard {...exampleData} />);

    expect(container).toMatchSnapshot();
  });

  it("snapshot a long card skeleton", () => {
    const { container } = render(<LongCard.Skeleton />);

    expect(container).toMatchSnapshot();
  });
});
