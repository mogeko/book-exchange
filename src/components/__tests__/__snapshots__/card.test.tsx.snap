// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshot a book card 1`] = `
<div>
  <div
    class="relative flex flex-col mb-4 w-28"
  >
    <figure
      class="flex justify-center items-center"
    >
      <a
        class="relative w-full h-[calc(7rem/2*3)]"
        href="/books/1"
      >
        <span
          style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
        >
          <img
            alt="This is an example book"
            data-nimg="fill"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </a>
    </figure>
    <div
      class="flex flex-col text-sm"
    >
      <a
        class="link no-underline text-primary hover:bg-primary hover:text-primary-content focus:bg-primary-focus focus:text-primary-content"
        href="/books/1"
      >
        <h2
          class="truncate"
        >
          This is an example book
        </h2>
      </a>
      <p
        class="truncate"
      >
        No One
      </p>
    </div>
  </div>
</div>
`;

exports[`Card snapshot a book skeleton card 1`] = `
<div>
  <div
    class="animate-pulse relative flex flex-col mb-4 w-28"
  >
    <figure
      class="flex justify-center items-center h-[calc(7rem/2*3)] rounded bg-base-300 mb-2"
    />
    <div
      class="flex flex-col"
    >
      <h2
        class="w-24 h-2 rounded bg-base-300 mb-2"
      />
      <p
        class="w-20 h-2 rounded bg-base-300 mb-2"
      />
    </div>
  </div>
</div>
`;
